core/src/subgraph/instance.rs

```diff
@@ -1,32 +1,47 @@
+use lazy_static::lazy_static;
+use std::env;
+use std::str::FromStr;
+
 use graph::prelude::{SubgraphInstance as SubgraphInstanceTrait, *};
 use graph::web3::types::{Log, Transaction};
 
+lazy_static! {
+    static ref MAX_DATA_SOURCES: Option<usize> = env::var("GRAPH_SUBGRAPH_MAX_DATA_SOURCES")
+        .ok()
+        .map(|s| usize::from_str(&s)
+            .unwrap_or_else(|_| panic!("failed to parse env var GRAPH_SUBGRAPH_MAX_DATA_SOURCES")));
+}
+
 pub struct SubgraphInstance<T>
 where
-    T: RuntimeHostBuilder,
+    T: RuntimeHostBuilder + Sync,
 {
     /// Runtime hosts, one for each data source mapping.
     ///
     /// The runtime hosts are created and added in the same order the
     /// data sources appear in the subgraph manifest. Incoming block
     /// stream events are processed by the mappings in this same order.
     hosts: Vec<Arc<T::Host>>,
+
+    /// Builder for runtime hosts.
+    host_builder: T,
+
+    /// Dynamic data sources.
+    ethereum_log_filter: EthereumLogFilter,
```

Comment:
Putting it on the builder feels too implicit. And since the builder is cloned for each subgraph, clearing the filter would be too hidden too, IMHO.

But I agree, this shouldn't be on the subgraph instance. I've simplified this now, so there's only _one_ log filter ever and it's on the `SubgraphState`. From there it is passed to the `BlockStreamBuilder` only when it is time to build a new stream. See the new commit in this PR:

https://github.com/graphprotocol/graph-node/commit/5ad7f4db6e9a969ba655e0cdcd864236600540e3

Reply:
Nice solution!